### PR TITLE
Semantic OrderView

### LINKED ISSUE #116: [Proposal] SemanticOrderView
# SemanticOrderView

* [x] Proposed
* [x] Prototype
* [x] Implementation
  * [x] iOS Support
  * [x] Android Support
  * [x] macOS Support
  * [x] Windows Support
* [x] Unit Tests
* [x] Sample
* [x] Documentation: https://github.com/MicrosoftDocs/CommunityToolkit/pull/236

## Summary
[summary]: #summary

Accessibility augmentation to set the order of VisualElements for screen readers

## Detailed Design
[design]: #detailed-design

### SemanticOrderView.shared.cs

```cs
public class SemanticOrderView : ContentView
{
  public static readonly BindableProperty ViewOrderProperty =
	  BindableProperty.Create(nameof(ViewOrder), typeof(IEnumerable), typeof(SemanticOrderView), Array.Empty<View>());
  
  public IEnumerable ViewOrder
  {
	  get => (IEnumerable)GetValue(ViewOrderProperty);
	  set => SetValue(ViewOrderProperty, value);
  }
  
  public SemanticOrderView()
  {
  }
}
```

## Usage Syntax
[usage]: #usage-syntax

<!-- Here is an example from CharactersValidationBehavior:

### XAML Usage

```xml
<StackLayout Margin="20">
    <Button Text="Element outside the Semantic View"></Button>
    <xct:SemanticOrderView x:Name="acv">
        <StackLayout>
            <Label x:Name="second" Text="Second" Margin="0,20" />
            <Button x:Name="third" Text="Third" Margin="0,20" />
            <Label x:Name="fourth" Text="Fourth" Margin="0,20" />
            <Button x:Name="fifth" Text="Fifth and last" Margin="0,20" />
            <Button x:Name="first" Text="First" Margin="0,20" />
        </StackLayout>
    </xct:SemanticOrderView>
</StackLayout>
```

### C# Usage

```cs
public partial class SemanticOrderViewPage : BasePage
{
  public SemanticOrderViewPage()
  {
    InitializeComponent();
    acv.ViewOrder = new List<View> { first, second, third, fourth, fifth };
  }
}
```

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

 

 ### Linked Issues ###
 

 - Fixes #116

 ### PR Checklist ###
 
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/236
 ### Additional information ###

### CODE REVIEW

File: samples/CommunityToolkit.Maui.Sample/Pages/Views/SemanticOrderView/SemanticOrderViewPage.xaml
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/Pages/Views/SemanticOrderView/SemanticOrderViewPage.xaml:12]: let's use VerticalStackLayout

File: src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.shared.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.shared.cs:43]:
```suggestion
	static void MapViewOrder(SemanticOrderViewHandler handler, ISemanticOrderView view)
```

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.android.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.android.cs:9]:
```suggestion
/// The native implementation of the <see href="SemanticOrderView"/> control.
```

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs:9]:
```suggestion
/// The native implementation of the <see href="SemanticOrderView"/> control.
```

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs:28]:
```suggestion
		if (result is not null)
```

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs:34]: These will be converted to an array, can we change the return type to be an `IEnumerable<NSObject>?` ? That way we don't allocate the List in the memory, just the array that will be consumed. Something like that

```csharp
	internal void UpdateViewOrder()
	{
		var result = GetAccessibilityElements();

		if (result.Any())
		{
			this.SetAccessibilityElements(NSArray.FromNSObjects(result.ToArray()));
		}
	}

	IEnumerable<NSObject> GetAccessibilityElements()
	{
		if (VirtualView is null)
		{
			yield break;
		}

		var viewOrder = VirtualView.ViewOrder;

		foreach (var view in viewOrder)
		{
			if (view.Handler is IPlatformViewHandler pvh &&
				pvh.PlatformView is not null)
			{
				yield return pvh.PlatformView;
			}
		}
	}
```

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.windows.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.windows.cs:11]:
```suggestion
/// The native implementation of the <see href="SemanticOrderView" /> control.
```

File: src/CommunityToolkit.Maui/Views/SemanticOrderView/SemanticOrderView.shared.cs
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Views/SemanticOrderView/SemanticOrderView.shared.cs:16]:
```suggestion
		BindableProperty.Create(nameof(ViewOrder), typeof(IEnumerable), typeof(SemanticOrderView), Enumerable.Empty<View>());
```

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.android.cs
Comment (PureWeen) [reply src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.android.cs:9]: Side note, I just copied these from the platform popup control so the MauiPopup platform view probably also needs these fixes

**Comment by PureWeen** [comment]
@pictos @VladislavAntonyuk 

the windows build failed but I don't know where/why/how it failed. 
I don't have access to the CI pipeline either

**Comment by pictos** [comment]
@PureWeen your branch is up to date with `main`? That was because our CI didn't found the tizen workload, but that should be solved already at `main` branch

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs
Comment (pictos) [inline src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs:28]: [From my tests](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRj1UwAJUCAWAbmLIoDoAZASygEdbdiCBOACgBEgHg3AIPuCAlFzwA3AIZhSAewBGAK1IBeUlACmAd1IBhJQFsADgBtdADwDy6/hLwBvPKQ+kAaizDAIcpY+Blo6EJaWMO6ewfp2YHC6itouBDDoMADMMJQAvni50ty48oo2oQDiusAAyma6XoEQugDO/KpqUng8AjYMJtDATl24AGZKYLpyAMYAFvylpNOkbKQ2zrie5Hz80yPd+ASZADxswAB8pFW19Y2WzW0mFtb26srqG26bniyjpO3qBg+PwBIIsEIsFphCIbLZbCgEUgqSZyADWXC20Q84xRcwWClIMhWUHeaiB4LiCSSsM8XzhPz+C1IAFJSOgVlDMDT6fCCIjUAB2QkYzz5XBig7EdlPKy2BxqVxY8iZUm6abAAD83l8/kCsVILgA5tVqKQWiaxZjvh5UCrWC1gKcoBdtQZ4olFEaTWaLaEAKJQCCmJJyFTWBh+izAACeTouTgYABUlPahiNckA===), this will never be null, but it will be an empty collection instead

**Comment by VladislavAntonyuk** [comment]
```
D:\a\1\s\src\CommunityToolkit.Maui.Core\Handlers\SemanticOrderView\SemanticOrderViewHandler.shared.cs(9,22): error CS0263: Partial declarations of 'SemanticOrderViewHandler' must not specify different base classes [D:\a\1\s\src\CommunityToolkit.Maui.Core\CommunityToolkit.Maui.Core.csproj::TargetFramework=net7.0-tizen]
D:\a\1\s\src\CommunityToolkit.Maui.Core\Handlers\SemanticOrderView\SemanticOrderViewHandler.shared.cs(14,78): error CS0311: The type 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' cannot be used as type parameter 'TViewHandler' in the generic type or method 'IPropertyMapper<TVirtualView, TViewHandler>'. There is no implicit reference conversion from 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' to 'Microsoft.Maui.IElementHandler'. [D:\a\1\s\src\CommunityToolkit.Maui.Core\CommunityToolkit.Maui.Core.csproj::TargetFramework=net7.0-tizen]
D:\a\1\s\src\CommunityToolkit.Maui.Core\Handlers\SemanticOrderView\SemanticOrderViewHandler.shared.cs(21,76): error CS0311: The type 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' cannot be used as type parameter 'TViewHandler' in the generic type or method 'CommandMapper<TVirtualView, TViewHandler>'. There is no implicit reference conversion from 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' to 'Microsoft.Maui.IElementHandler'. [D:\a\1\s\src\CommunityToolkit.Maui.Core\CommunityToolkit.Maui.Core.csproj::TargetFramework=net7.0-tizen]

```

File: src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs
Comment (PureWeen) [reply src/CommunityToolkit.Maui.Core/Views/SemanticOrderView/MauiSemanticOrderView.ios.cs]: 👍

**Comment by PureWeen** [comment]
> Tizen build fails.
> 
> ```
> D:\a\1\s\src\CommunityToolkit.Maui.Core\Handlers\SemanticOrderView\SemanticOrderViewHandler.shared.cs(9,22): error CS0263: Partial declarations of 'SemanticOrderViewHandler' must not specify different base classes [D:\a\1\s\src\CommunityToolkit.Maui.Core\CommunityToolkit.Maui.Core.csproj::TargetFramework=net7.0-tizen]
> D:\a\1\s\src\CommunityToolkit.Maui.Core\Handlers\SemanticOrderView\SemanticOrderViewHandler.shared.cs(14,78): error CS0311: The type 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' cannot be used as type parameter 'TViewHandler' in the generic type or method 'IPropertyMapper<TVirtualView, TViewHandler>'. There is no implicit reference conversion from 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' to 'Microsoft.Maui.IElementHandler'. [D:\a\1\s\src\CommunityToolkit.Maui.Core\CommunityToolkit.Maui.Core.csproj::TargetFramework=net7.0-tizen]
> D:\a\1\s\src\CommunityToolkit.Maui.Core\Handlers\SemanticOrderView\SemanticOrderViewHandler.shared.cs(21,76): error CS0311: The type 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' cannot be used as type parameter 'TViewHandler' in the generic type or method 'CommandMapper<TVirtualView, TViewHandler>'. There is no implicit reference conversion from 'CommunityToolkit.Maui.Core.Handlers.SemanticOrderViewHandler' to 'Microsoft.Maui.IElementHandler'. [D:\a\1\s\src\CommunityToolkit.Maui.Core\CommunityToolkit.Maui.Core.csproj::TargetFramework=net7.0-tizen]
> ```

🤞

**Comment by JoonghyunCho** [comment]
Sorry for the Tizen workload issue. Looks like it is ok now :)

File: src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs
Comment (JoonghyunCho) [inline src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs:9]: Should be `CreatePlatformView()`? with `using Microsoft.Maui.Platform;`.

Comment (JoonghyunCho) [reply src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs:9]: Later I will check the Tizen implementation for `MauiSemanticOrderView`!

Comment (PureWeen) [reply src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs]: Thank you @JoonghyunCho !

**Review by pictos — APPROVED** [review]

File: src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs:12]: @JoonghyunCho - are you able to add the Tizen implementation before me merge this PR?

**Review by brminnick — CHANGES_REQUESTED** [review]
Thanks Shane! Let's see if @JoonghyunCho is available to implement this for Tizen before merging this PR 👍

File: src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs
Comment (JoonghyunCho) [reply src/CommunityToolkit.Maui.Core/Handlers/SemanticOrderView/SemanticOrderViewHandler.tizen.cs]: Oops. Just found the message today waiting for Tizen 😭. Sorry for little delay. 
Good news! Just finished implementing Tizen implementation 😄

**Review by jfversluis — DISMISSED** [review]

**Review by pictos — DISMISSED** [review]

**Comment by brminnick** [comment]
Thanks @PureWeen! Are you able to submit a PR to our docs repo today? We are releasing v5.0.0 tomorrow, and I'd love to include this in the release! But we can't merge it until we have a Docs PR opened: https://github.com/MicrosoftDocs/CommunityToolkit/pull/

@bijington or @pictos - are you by chance available to help with the docs?

File: samples/CommunityToolkit.Maui.Sample/Pages/Views/SemanticOrderView/SemanticOrderViewPage.xaml.cs
Comment (brminnick) [inline samples/CommunityToolkit.Maui.Sample/Pages/Views/SemanticOrderView/SemanticOrderViewPage.xaml.cs:11]: What is the recommended way for users to implement `ViewOrder`? 

If it's with a ViewModel bindings, let's move this to `SemanticOrderPageViewModel`.
If it's in XAML, let's move this to XAML.

File: src/CommunityToolkit.Maui/Views/SemanticOrderView/SemanticOrderView.shared.cs
Comment (brminnick) [inline src/CommunityToolkit.Maui/Views/SemanticOrderView/SemanticOrderView.shared.cs:19]: How come this `IEnumerable` doesn't have a type?

Should it be `IEnumerable<View>` or `IEnumerable<IView>`?

**Review by brminnick — COMMENTED** [review]
Thanks Shane! 

I'd love a quick update to the Sample App page to provide more context and guide the user on what they are learning/experiencing by visiting this page in the sample app.

I'm also curious why `CommunityToolkit.Maui.Views.SemanticOrderView.ViewOrder` is using `IEnumerable` instead of `IEnumerable<View>` or `IEnumerable<IView>`?

**Comment by bijington** [comment]
I'm on the docs case right now!

**Review by brminnick — APPROVED** [review]

**Comment by PureWeen** [comment]
Thank you for finishing this one up!
I just got back into the office today and just caught up to the requests here. 

!!!! Thank You !!!!

**Pull request merged**
